packages/block-editor/src/components/html-element-select-control/index.js

```diff
@@ -0,0 +1,135 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { SelectControl, Notice } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import { store as blockEditorStore } from '../../store';
+import { htmlElementMessages } from './messages';
+
+/**
+ * Checks if a block has a specific HTML element tag.
+ *
+ * @param {Object} block Block object to check.
+ * @param {string} tag   HTML tag to look for.
+ * @return {boolean}     Whether the block has the specified HTML element.
+ */
+const blockHasHtmlTag = ( block, tag ) => {
+	if ( ! block ) {
+		return false;
+	}
+
+	if ( block.attributes && block.attributes.tagName === tag ) {
+		return true;
+	}
+
+	if ( block.innerBlocks && block.innerBlocks.length ) {
+		return block.innerBlocks.some( ( innerBlock ) =>
+			blockHasHtmlTag( innerBlock, tag )
+		);
+	}
+
+	return false;
+};
+
+/**
+ * Renders a SelectControl for choosing HTML elements with validation
+ * to prevent duplicate <main> elements.
+ *
+ * @param {Object}   props                 Component props.
+ * @param {string}   props.tagName         The current HTML tag name.
+ * @param {Function} props.onChange        Function to call when the tag is changed.
+ * @param {string}   props.currentClientId The client ID of the current block.
+ * @param {Array}    props.options         SelectControl options (optional).
+ *
+ * @return {Component} The HTML element select control with validation.
+ */
+export default function HTMLElementSelectControl( {
+	tagName,
+	onChange,
+	currentClientId,
+	options = [
+		{ label: __( 'Default (<div>)' ), value: 'div' },
+		{ label: '<header>', value: 'header' },
+		{ label: '<main>', value: 'main' },
+		{ label: '<section>', value: 'section' },
+		{ label: '<article>', value: 'article' },
+		{ label: '<aside>', value: 'aside' },
+		{ label: '<footer>', value: 'footer' },
+	],
+} ) {
+	const { hasMultipleMainElements, hasMainElementElsewhere } = useSelect(
+		( select ) => {
+			const { getBlocks, getBlock } = select( blockEditorStore );
+			const allBlocks = getBlocks();
+			const currentBlock = getBlock( currentClientId );
```

Comment:
Suggestion (blocking): The `getBlocks` selector won't work 100% when the editor is rendered in `template-locked` mode. Try toggling "Show Template" when testing the feature.

We can probably use the `getClientIdsWithDescendants` + `getBlockAttributes` combo.